### PR TITLE
[AIRFLOW-7065] Add optional propagation of task std streams to console

### PR DESCRIPTION
**Pull request by c-wilson**
Both `stdout` and `stderr` are captured during Task execution to allow handling by custom log writers. While this is great, there are cases where forwarding task execution output to system streams (console) during task runtime is advantageous (ie for capture by docker logging drivers). Adding a `LogHandler` that wrote to stdout or stderr would result in an infinite loop of message forwarding `stdout > StreamLogWriter > logger > stdout > StreamLogWriter > ...`

This adds a configuration flag `['logging']['task_log_to_console']` which enables forwarding of task stream messages to the console when set. This is implemented via a new context manager that adds existing system stream targets as targets for the `StreamLogWriter`. There is some implementation of log target replacement that is essentially duplicated from the stdlib, but I could not think of a better way to do this without a subclassing death-spiral. It has decent tests, and I do not think that this will be difficult to keep unbroken.

---
Issue link: [AIRFLOW-7065](https://issues.apache.org/jira/browse/AIRFLOW-7065)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by potiuk** [comment]
It's a very welcome change. Very useful for debugging. @mik-laj @ashb -> I know you were looking at the logging recently and know definitely more than me on that subject. Maybe you can take a look?

**Comment by ashb** [comment]
Isn't there a CLI flag that already triggers this? `--interactive`?

**Comment by ashb** [comment]
If we are making this change then I would like you to also look at the "write_to_stdout" path of the Elasticsearch log handler -- can we tidy up that path.

File: airflow/utils/log/logging_mixin.py [hunk 1]
```diff
@@ -58,21 +61,23 @@ def _set_context(self, context):
             set_context(self.log, context)
 
 
-# TODO: Formally inherit from io.IOBase
```
Comment (ashb) [inline airflow/utils/log/logging_mixin.py]: Why did you remove this todo? It should still be done in the future, no?

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -141,8 +140,8 @@ def _run_file_processor(result_channel,
 
         try:
             # redirect stdout/stderr to log
-            with redirect_stdout(StreamLogWriter(log, logging.INFO)),\
-                    redirect_stderr(StreamLogWriter(log, logging.WARN)):
+            with StdoutToLog(log, logging.INFO), \
+                    StderrToLog(log, logging.WARN):
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: Is there any need to make this change? What does this achieve?

Comment (c-wilson) [reply airflow/jobs/scheduler_job.py]: Just made this change for consistency, but you are correct that it is not necessary. Probably not actually helpful either on second look.

File: airflow/utils/log/logging_mixin.py (hunk 1 above)
Comment (c-wilson) [reply airflow/utils/log/logging_mixin.py]: My mistake, will revert

**Comment by c-wilson** [comment]
I can certainly look at the ES log path. The benefit of the ES handler is that it prints _formatted_ messages to console, which seems helpful. The downside is that it doesn’t stream to console in real time but reads from the log file on task completion. It also is hard coded to sys dunder streams. 

I wonder if a better implementation than my current one  would be to add a LogHandler that writes to the existing stream target instead of writing directly from the `StreamLogWriter`. This would avoid the feedback loop while preserving logging behavior. I can explore this.

**Comment by ashb** [comment]
To be clear I'm not saying the ES log handler is right for your use case, just to see if the code there can be improved in light of this PR

File: airflow/utils/log/logging_mixin.py [hunk 3]
```diff
@@ -118,6 +129,53 @@ def isatty(self):
...
+
+        setattr(sys, self.stream_to_replace, self._replacement_stream)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Puts back existing stream target"""
+        self._replacement_stream.flush()
+        setattr(sys, self.stream_to_replace, self._existing_stream_target.pop())
+
+
+class StdoutToLog(_StreamToLogRedirector):
```
Comment (mik-laj) [inline airflow/utils/log/logging_mixin.py:170]: Can we use ``contextlib.redirect_stdout`` here?

Comment (mik-laj) [inline airflow/utils/log/logging_mixin.py:170]:
```suggestion
class stdout_to_log(_StreamToLogRedirector):
```
Python code convention suggests that context managers should be in snake_case. CamelCase is used when we want to emphasize that it is a class, but it is not important here.
https://github.com/python/cpython/blob/3.8/Lib/contextlib.py#L336

**Comment by c-wilson** [comment]
Hi @ashb I made some trivial cleanup changes to the ESTaskHandler but in reality I cannot find a better way to implement its functionality with the current (slightly deep) logging hierarchical pattern:

```airflow.LogStreamHandler > logging.logger > airflow.TaskLogHandler > logger.LogHandler > stream```

In this flow, the ESTaskHandler really is just injecting a JSON formatter before the standard out stream as far as I can see. Seems that the sin of the ESTaskHandler is overloading the stdout and file writing paths in the same class which is a bit surprising.

It is maybe a thankless and non-urgent task, but it may be worth a reevaluation of how the logging environment is composed at a broader level. When you enter a task context, factory functions could compose real stdlib `logging.Handler` instances instead of the current subclassing pattern. These could be composed such that they are pointed at desired streams (file or console) and have the desired formatters and could be attached to the logger via logger.addHandler. I'm not sure that this _actually_ cleaner/better, the current implementation obviously works, and I'm not sure how important it is that it be maximally readable/extensible.

File: airflow/utils/log/logging_mixin.py (hunk 3 above)
Comment (c-wilson) [reply airflow/utils/log/logging_mixin.py:170]: It's possible, but I don't know if the result would be cleaner. Maybe you're right. This just serves as a compact way to pass the right stream to the StreamLogHandler when propagation is desired. This _could_ live in the task_command.py as if statements if we want.

**Pull request closed without merging**